File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    private boolean isPossible(int []weights, int cap, int days){
        int dayCount = 1;
        int wt = 0;
        for(int el : weights){
            int currentCap = wt + el;
            if(currentCap <= cap) wt = currentCap;
            else{
                dayCount++;
                wt = el;
            }
        }
        
        return dayCount <= days;
    }
    
    public int shipWithinDays(int[] weights, int days) {
        int si, ei = 0;
        int max = Integer.MIN_VALUE;
        for(int el : weights){
            max = Math.max(max, el);
            ei += el;
        }
        // Assigining the max value to si:
        si = max;
        
        int pans = -1;
        while(si <= ei){
            int mid = si + (ei - si)/2;
            if(isPossible(weights, mid, days)){
                pans = mid;
                ei = mid -1;
            }
            else si = mid +1;
        }
        
        return pans;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(weights, days) {
    let isPossible = (mid, d) =>{
        let day = 1, wt = 0;
        weights.forEach(el =>{
            let currWt = wt + el;
            if(currWt <= mid) wt = currWt;
            else{
                wt = el;
                day++;
            }
        })
        
        return day <= d;
    }
    // Edge Cases:
    if(days > weights.length) return -1;
    
    let max = Number.MIN_SAFE_INTEGER, sum = 0;
    weights.forEach(el => {
        max = Math.max(max, el);
        sum += el;
    });
    
    let low = max, high = sum;
    let pans = 0;
    while(low<=high){
        let mid = Math.floor(low + (high - low)/2); // Weight
        if(isPossible(mid, days)){
            // Find the least possible weight:
            pans = mid;
            high = mid -1;
        }
        else{
            // Find the least possible weight:
            low = mid +1;
        }
    }
    return pans;
};